pkg/apiserver/registry/system/supportbundle/rest.go

```diff
@@ -0,0 +1,417 @@
+	sparkDriverLabel     = "spark-role=driver"
+	sparkExecLabel       = "spark-role=executor"
+	sparkOperatorLabel   = "app.kubernetes.io/name=spark-operator"
+	zookeeperLabel       = "app=zookeeper"
+)
+
+var (
+	// Declared as variables for testing.
+	defaultFS        = afero.NewOsFs()
+	newManagerDumper = support.NewManagerDumper
+)
+
+// NewSupportBundleStorage creates a support bundle storage for working on theia manager.
+func NewSupportBundleStorage(config *clientrest.Config, clientset kubernetes.Interface) Storage {
+	namespace := env.GetTheiaNamespace()
+	cpConfig := clientrest.CopyConfig(config)
+	cpConfig.APIPath = "/api"
+	cpConfig.GroupVersion = &schema.GroupVersion{Version: "v1"}
+	cpConfig.NegotiatedSerializer = serializer.WithoutConversionCodecFactory{CodecFactory: clientscheme.Codecs}
+
+	bundle := &supportBundleREST{
+		cache: &systemv1alpha1.SupportBundle{
+			ObjectMeta: metav1.ObjectMeta{Name: bundleName},
+			Status:     systemv1alpha1.SupportBundleStatusNone,
+		},
+		namespace:  namespace,
+		restConfig: cpConfig,
+		clientSet:  clientset,
+	}
+	return Storage{
+		SupportBundle: bundle,
+		Download:      &downloadREST{supportBundle: bundle},
+	}
+}
+
+// Storage contains REST resources for support bundle, including status query and download.
+type Storage struct {
+	SupportBundle *supportBundleREST
+	Download      *downloadREST
+}
+
+var (
+	_ rest.Scoper          = &supportBundleREST{}
+	_ rest.Getter          = &supportBundleREST{}
+	_ rest.Creater         = &supportBundleREST{}
+	_ rest.GracefulDeleter = &supportBundleREST{}
+)
+
+// supportBundleREST implements REST interfaces for bundle status querying.
+type supportBundleREST struct {
+	statusLocker sync.RWMutex
+	cancelFunc   context.CancelFunc
+	cache        *systemv1alpha1.SupportBundle
+
+	namespace  string
+	restConfig *clientrest.Config
+	clientSet  kubernetes.Interface
+}
+
+// Create triggers a bundle generation. Returns metav1.Status if there is any error,
+// otherwise it returns the SupportBundle.
+func (r *supportBundleREST) Create(ctx context.Context, obj runtime.Object, _ rest.ValidateObjectFunc, _ *metav1.CreateOptions) (runtime.Object, error) {
+	requestBundle := obj.(*systemv1alpha1.SupportBundle)
+	if requestBundle.Name != bundleName {
+		return nil, errors.NewForbidden(systemv1alpha1.SupportBundleResource.GroupResource(), requestBundle.Name,
+			fmt.Errorf("only resource name \"%s\" is allowed", bundleName))
+	}
+	r.statusLocker.Lock()
+	defer r.statusLocker.Unlock()
+
+	if r.cancelFunc != nil {
+		r.cancelFunc()
+	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	r.cache = &systemv1alpha1.SupportBundle{
+		ObjectMeta: metav1.ObjectMeta{Name: requestBundle.Name},
+		Since:      requestBundle.Since,
+		Status:     systemv1alpha1.SupportBundleStatusCollecting,
+	}
+	r.cancelFunc = cancelFunc
+	go func(since string) {
+		var err error
+		var b *systemv1alpha1.SupportBundle
+		b, err = r.collectController(ctx, since)
+		func() {
```

Comment:
Just curious, why do we need an anonymous function here?

Reply:
This is mainly to make the status update part (protected by statusLocker) explicit in its own scope / func call, and to utilize lock() / defer unlock() semantic. Otherwise we need to unlock based on different conditions, which could be less maintainable and prone to error.